includes/gateways/paypal/ipn.php

```diff
@@ -0,0 +1,400 @@
+							/* Translators: %s - The collected outstanding balance of the subscription */
+							$subscription->add_note( sprintf( __( 'Outstanding subscription balance of %s collected successfully.', 'easy-digital-downloads' ), $amount ) );
+						}
+					} else {
+						ipn_debug_log( 'subscription ' . $subscription->id . ': renewal payment creation appeared to fail.' );
+					}
+
+					die( 'Subscription payment successful' );
+
+					break;
+
+				case 'recurring_payment_profile_cancel':
+				case 'recurring_payment_suspended':
+				case 'recurring_payment_suspended_due_to_max_failed_payment':
+					if ( 'cancelled' === $subscription->status ) {
+						ipn_debug_log( 'Subscription ID ' . $subscription->id . ' arlready cancelled.' );
+						return;
+					}
+
+					$subscription->cancel();
+					ipn_debug_log( 'subscription ' . $subscription->id . ': subscription cancelled.' );
+
+
+					die( 'Subscription cancelled' );
+
+					break;
+
+				case 'recurring_payment_failed':
+					if ( 'failing' === $subscription->status ) {
+						ipn_debug_log( 'Subscription ID ' . $subscription->id . ' arlready failing.' );
+						return;
+					}
+
+					$subscription->failing();
+					ipn_debug_log( 'subscription ' . $subscription->id . ': subscription failing.' );
+					do_action( 'edd_recurring_payment_failed', $subscription );
+
+					break;
+
+				case 'recurring_payment_expired':
+					if ( 'completed' === $subscription->status ) {
+						ipn_debug_log( 'Subscription ID ' . $subscription->id . ' arlready completed.' );
+						return;
+					}
+
+					$subscription->complete();
+					ipn_debug_log( 'subscription ' . $subscription->id . ': subscription completed.' );
+
+					die( 'Subscription completed' );
+					break;
+
+			endswitch;
+		}
+
+		// We've processed recurring, now let's handle non-recurring IPNs.
+		$order_id = 0;
+
+		if ( ! empty( $posted['parent_txn_id'] ) ) {
+			$order_id = edd_get_order_id_from_transaction_id( $posted['parent_txn_id'] );
+		}
+
+		/**
+		 * This section of the IPN should only affect processing refunds or returns on orders made previous, not new
+		 * orders, so we can just look for the parent_txn_id, and if it's not here, bail as this is a new order that
+		 * should be handeled with webhooks.
+		 */
+		if ( empty( $order_id ) ) {
+			ipn_debug_log( 'IPN Track ID ' . $posted['ipn_track_id'] . ' does not need to be processed as it does not belong to an existing record.' );
+			return;
+		}
+
+		$order = edd_get_order( $order_id );
+		if ( 'paypal_commerce' !== $order->gateway ) {
+			ipn_debug_log( 'Order ' . $order_id . ' was not with PayPal Commerce' );
+			return;
+		}
+
+		if ( 'refunded' === $payment_status || 'reversed' === $payment_status ) {
+			$order = edd_get_order( $order_id );
+			if ( 'refunded' === $order->status ) {
+				ipn_debug_log( 'Order ' . $order_id . ' is already refunded' );
+			}
+
+			$transaction_exists = edd_get_order_transaction_by( 'transaction_id', $order->get_transaction_id() );
```

Comment:
I agree with @mihaijoldis here as well--the code as it currently is is redundant because it's getting the order that was already gotten in line 312, but that's the parent order, not the refund order that we're listening for:
```suggestion
			$transaction_exists = edd_get_order_transaction_by( 'transaction_id', $transaction_id );
```